Display app names in update page for app updates

PVince81 opened the pull request:
Whenever the update page is displayed for apps, show app names instead
of the core update text.

Fixes https://github.com/owncloud/core/issues/16272
- [x] TODO: add unit test
- [x] To discuss: should we move `getIncompatibleApps` and `getAppsNeedingUpgrade` to the `IAppManager` interface ? I'd rather keep them private but "apps.php" is horribly static... @icewind1991 @MorrisJobke @nickvergessen @oparoz what do you think ?

Commit:
Display app names in update page for app updates

Whenever the update page is displayed for apps, show app names instead
of the core update text.

PVince81 commented:
- [x] TEST: core update: shows core update page
- [x] TEST: app update alone: shows app update page
- [x] TEST: core + app update: shows core update page

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/30ef3999-8547-4862-904e-e12c580ca659) was created.

PVince81 commented:
- [x] TEST: auto-disable third party apps on core update still works

PVince81 commented:
Setting to "To review" because I'd like to get comments :smile:

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/78edc0fa-cdc4-4114-be60-2ab4e88e8810) was created.

PVince81 commented:
Fixes https://github.com/owncloud/core/issues/17437 too, @jancborchardt

PVince81 commented:
@jancborchardt I'll revert and fix separately, it's a different piece of code.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/29bf367c-9482-45c1-be81-626011c70a7d) was created.

PVince81 commented:
Ok looks like nobody objected the last 1.5 months so I'll keep these methods static.

Time for unit tests then.

PVince81 commented:
Basically it looks like this:
![versions-app-update](https://cloud.githubusercontent.com/assets/277525/9356404/cb5ba5c6-4680-11e5-88ab-39307d2cc085.png)

@jancborchardt

oparoz commented:
Regarding the static methods. I don't think app devs would have a use for it, unless they write an app manager?

Regarding the design, I think it would be nice if apps could be shown using bullet points. It makes it clearer what is info-text and what is an app.

PVince81 commented:
I checked the code and remembered why I can't move them easily to the app manager: the app manager is missing methods like "isAppCompatible" and "shouldUpgrade". I don't think we want these on the app manager either.

For the bullet points I'll let @jancborchardt decide.
I think until now we never had bullet points on this page for lists. Another list is the one that shows you what apps are getting disabled during a core update.

oparoz commented:
> Another list is the one that shows you what apps are getting disabled during a core update.

Ah, yes, that needs fixing as well because it only shows the folder name iirc.

PVince81 commented:
@oparoz please raise another ticket for that. Thanks.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/095279cb-75f9-4a85-9e0c-debbdbcb2b1a) was created.

PVince81 commented:
Done.

I moved `getIncompatibleApps()` and `getAppsNeedingUpgrade()` to the `AppManager` class.
I also added `AppManager->getAppInfo()` that calls the old API to make it mockable for unit tests.

Please review @icewind1991 @nickvergessen @MorrisJobke @Xenopathic @oparoz

Xenopathic commented on core/templates/update.admin.php:
```diff
@@ -1,7 +1,20 @@
 <div class="update" data-productname="<?php p($_['productName']) ?>" data-version="<?php p($_['version']) ?>">
 	<div class="updateOverview">
+		<?php if ($_['isAppsOnlyUpgrade']) { ?>
```
Do app upgrades not occur at the same time as core upgrades? If they do, then we should surely display 'The following apps will be updated' alongside 'ownCloud will be updated'?

Xenopathic replied:
As far as I can tell (without doing an actual upgrade), we will see:

> ownCloud will be updated to version x
> 
> External file storage (files_external)
> My awesome app (xenopathic_is_awesome)

IMHO we still need to display 'The following apps will be updated' under 'ownCloud will be updated'

PVince81 replied:
Yeah, I also thought about this.
I'll try and add them to both pages.

Xenopathic commented on core/templates/update.admin.php:
```diff
@@ -17,9 +30,11 @@
 			<?php p($l->t('The theme %s has been disabled.', array($_['oldTheme']))) ?>
 		</div>
 		<?php } ?>
+		<?php if (!$_['isAppsOnlyUpgrade']) { ?>
```
When updating apps, there is just as much chance for a DB schema upgrade to go wrong as there is in core. Same goes for the data folder. IMHO we should display this warning always

PVince81 replied:
Hmm, right

Xenopathic commented on lib/base.php:
```diff
@@ -375,6 +355,40 @@ public static function checkUpgrade($showTemplate = true) {
+	private static function printUpgradePage() {
+		$systemConfig = \OC::$server->getSystemConfig();
+		$oldTheme = $systemConfig->getValue('theme');
+		$systemConfig->setValue('theme', '');
```
This is a thing?! :see_no_evil:

PVince81 replied:
Possibly... this code was just moved here, I didn't write it.

MorrisJobke replied:
Yes. Please leave it. It caused a lot of trouble in the past.

PVince81 commented on lib/base.php:
```diff
@@ -375,6 +355,40 @@ public static function checkUpgrade($showTemplate = true) {
+		$tmpl->assign('version', OC_Util::getVersionString());
+
+		// if not a core upgrade, then it's apps upgrade
+		if (version_compare($currentVersion, $installedVersion, '=')) {
```
This is the part I added

PVince81 commented:
- [ ] TODO: add list of apps to upgrade on both core and app upgrade modes

PVince81 commented:
- [ ] TODO: keep warning about backups in both modes

jancborchardt commented:
I agree with @oparoz that it would be nicer to use a list to show the apps. That way they are easily readable and also not center-aligned.

PVince81 commented:
It is already a ul/li list, so I'll switch its CSS style to use bullets instead.
I assume we'll want this for the list of disabled apps too.

- [ ] TODO: add bullets to CSS for "ul" in update page

Commit:
Additions to update page

Apps to update and to disable will always be shown.
Main title changes only when apps need updated, not core.
Added bullet style.
Exclude incompatible apps from updated apps list.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/owncloud/core/inspections/8e535bf8-5a8e-486a-9830-3511b6638f78) was created.

PVince81 commented:
Done.
I also changed the wording a bit:
![versions-app-update1](https://cloud.githubusercontent.com/assets/277525/9379924/f458b944-472c-11e5-8cd4-e6ed162046f2.png)

Note that the main title changes if a core update is required too.

Also fixed:
- [x] BUG: if an app needing update is also incompatible, don't show it in the update list

Xenopathic commented:
I'm not 100% sure on the 'Apps update required' wording, but I can't think of any alternatives, and the code looks good :+1:

MorrisJobke commented:
* [ ] disabled 3rdparty apps aren't listed

PVince81 commented:
@MorrisJobke they are listed after the update. Something to be done separately because it might require changing things in the update order.

PVince81 commented:
Note: the original purpose of this PR is to solve the confusion that people have between core update and app update ("why does my ownCloud suddenly wants to update again ?")

MorrisJobke commented:
Then it is working just fine :+1:

jancborchardt commented on core/templates/update.admin.php:
```diff
@@ -1,12 +1,26 @@
 <div class="update" data-productname="<?php p($_['productName']) ?>" data-version="<?php p($_['version']) ?>">
 	<div class="updateOverview">
+		<?php if ($_['isAppsOnlyUpgrade']) { ?>
+		<h2 class="title bold"><?php p($l->t('Apps update required.')); ?></h2>
```
This should say »App update required« (App without s, no dot at the end.)

MorrisJobke replied:
https://github.com/owncloud/core/pull/18452

jancborchardt replied:
Pull request at https://github.com/owncloud/core/pull/18453
